### PR TITLE
github actions from v3 to v4

### LINKED ISSUE #10993: stop using v3 of download-artifact and upload-artifact
From https://github.blog/changelog/2024-04-16-deprecation-notice-v3-of-the-artifact-actions/

"Starting November 30, 2024, GitHub Actions customers will no longer be able to use v3 of [actions/upload-artifact](https://github.com/actions/upload-artifact) or [actions/download-artifact](https://github.com/actions/download-artifact). Customers should update workflows to begin using [v4 of the artifact actions](https://github.blog/2024-02-12-get-started-with-v4-of-github-actions-artifacts/) as soon as possible. While v4 of the artifact actions improves upload and download speeds by up to 98% and includes several new features, there are [key differences](https://github.com/actions/upload-artifact?tab=readme-ov-file#breaking-changes) from previous versions that may require updates to your workflows. Please see [the documentation](https://github.com/actions/upload-artifact/blob/main/docs/MIGRATION.md) in the project repositories for guidance on how to migrate your workflows."

We tried to upgrade in this PR but saw failures:

- #10822

We should try again, soon. And we should compare notes with the frontend:

- https://github.com/IQSS/dataverse-frontend/issues/537

### PR DESCRIPTION
**Pull request by stevenwinship**
**What this PR does / why we need it**: Starting November 30, 2024, GitHub Actions customers will no longer be able to use v3 of [actions/upload-artifact](https://github.com/actions/upload-artifact) or [actions/download-artifact](https://github.com/actions/download-artifact).

**Which issue(s) this PR closes**:https://github.com/IQSS/dataverse/issues/10993

- Closes #10993

**Special notes for your reviewer**: I chose to update all the V3 to V4 actions in the yml files. Only the upload and download were required but I felt it's better to not mix/load multiple library versions if possible.
Due to a bug in download-artifact@V4 the version was bumped to 4.1.7 to include the fix
As mentioned in an issue that there was a bug in download V4 that was fixed in 4.1.7 https://github.com/actions/upload-artifact/issues/478
Update: v4 now gets the latest so download-artifact can now use v4 to get the fix in 4.1.7 (latest is v4.1.8 as of this PR)

**Suggestions on how to test this**: I've made temporary changes to various files to force most of the actions to run. Obviously any release specific actions were not run. I think the best way to test it is to merge it and monitor the build system. 

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

**Is there a release notes update needed for this change?**:

**Additional documentation**:

### CODE REVIEW

**Comment by pdurbin** [comment]
@stevenwinship this pull request looks fine if we want to do the same thing as last time... merge it, see if anything breaks, and revert it if necessary.

I see 15 commits. Did you do some testing? If so, can you please write a little about it?

"Suggestions on how to test this" is empty. Is this because we plan to just merge it and see if anything breaks?

**Comment by stevenwinship** [comment]
The reason for so many commits was indeed testing. I wanted to test with updating all iml files to V4 but ran into some issues so I ended up limiting the changes to the 2 files that were failing. I also tested with java changes and doc changes to force all the tests to run.
As for "How To Test", yes merging and monitoring would be the best way to verify it beyond what I've already done.

File: .github/workflows/deploy_beta_testing.yml
Comment (pdurbin) [inline .github/workflows/deploy_beta_testing.yml:48]: Is there a reason why "upload artifact" is set to "v4" but "download-artifact" is a specific version (v4.1.7)? Should we put a comment above the 4.1.7 lines to explain why it's pinned so specifically?

There's one case here and two more in the other file.

**Review by pdurbin — COMMENTED** [review]
@stevenwinship thanks for the explanation. Here's one more question before I approve this.

File: .github/workflows/deploy_beta_testing.yml
Comment (stevenwinship) [reply .github/workflows/deploy_beta_testing.yml:48]: As mentioned in an issue that there was a bug in download V4 that was fixed in 4.1.7 https://github.com/actions/upload-artifact/issues/478

Comment (stevenwinship) [reply .github/workflows/deploy_beta_testing.yml:49]: added comment

**Review by pdurbin — APPROVED** [review]
As before, we'll merge this and check if it breaks anything and revert if necessary.

**Comment by pdurbin** [comment]
Also, I assumed "v4" meant "the latest v4 but it doesn't. See https://github.com/orgs/community/discussions/25248

That is, v4 means https://github.com/actions/download-artifact/releases/tag/v4

v4 does not meant the latest, which as of this writing, is https://github.com/actions/download-artifact/releases/tag/v4.1.8

**Review by pdurbin — APPROVED** [review]
Looks good. As discussed at https://github.com/IQSS/dataverse-frontend/pull/550#issuecomment-2485837059 we've learned that "v4" really does mean "the latest 4" as I initially assumed.

Approved.

Let's revert this if it breaks anything.

**Comment by ofahimIQSS** [comment]
Merging PR - monitoring github actions

**Comment by pdurbin** [comment]
https://github.com/IQSS/dataverse/actions/workflows/deploy_beta_testing.yml is failing.

**Pull request merged**
